### PR TITLE
23_hotfix: WdlOptionalTypes as valid runtime attrs closes #1737

### LINKED ISSUE #1737: Configurable backend: Optional runtime attributes broke in 23
The following config used to work (0.22):
```
    PBS {
      actor-factory = "cromwell.backend.impl.sfs.config.ConfigBackendLifecycleActorFactory"
      config {
        runtime-attributes = """
        Int memory_mb = 1000
        String  pbs_cpu = "1"
        String? pbs_email
        String? pbs_queue
        String  pbs_walltime = "1:00:00"
        """
      }
    }
```

Note the specification of `pbs_email` and `pbs_queue` as Optional String type

On upgrading to 23 and starting the server, the process hangs and this appears in the logs:
```
[ERROR] [12/05/2016 13:28:57.994] [cromwell-system-akka.dispatchers.engine-dispatcher-30] [akka://cromwell-system/user/SingleWorkflowRunnerActor/WorkflowManagerActor/WorkflowActor-8def86a5-13b1-4dc4-9cdf-d0ae2eedd7c9/WorkflowInitializationActor-8def86a5-13b1-4dc4-9cdf-d0ae2eedd7c9/PBS] Unsupported config runtime attribute WdlOptionalType(WdlStringType) pbs_email
java.lang.RuntimeException: Unsupported config runtime attribute WdlOptionalType(WdlStringType) pbs_email
	at cromwell.backend.impl.sfs.config.DeclarationValidation$.fromDeclaration(DeclarationValidation.scala:41)
	at cromwell.backend.impl.sfs.config.DeclarationValidation$$anonfun$fromDeclarations$1.apply(DeclarationValidation.scala:17)
	at cromwell.backend.impl.sfs.config.DeclarationValidation$$anonfun$fromDeclarations$1.apply(DeclarationValidation.scala:17)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at cromwell.backend.impl.sfs.config.DeclarationValidation$.fromDeclarations(DeclarationValidation.scala:17)
	at cromwell.backend.impl.sfs.config.ConfigInitializationActor.declarationValidations$lzycompute(ConfigInitializationActor.scala:38)
	at cromwell.backend.impl.sfs.config.ConfigInitializationActor.declarationValidations(ConfigInitializationActor.scala:37)
	at cromwell.backend.impl.sfs.config.ConfigInitializationActor.runtimeAttributesBuilder$lzycompute(ConfigInitializationActor.scala:47)
	at cromwell.backend.impl.sfs.config.ConfigInitializationActor.runtimeAttributesBuilder(ConfigInitializationActor.scala:46)
	at cromwell.backend.sfs.SharedFileSystemInitializationActor.coerceDefaultRuntimeAttributes(SharedFileSystemInitializationActor.scala:90)
	at cromwell.backend.BackendWorkflowInitializationActor$class.validateRuntimeAttributes(BackendWorkflowInitializationActor.scala:138)
	at cromwell.backend.BackendWorkflowInitializationActor$class.initSequence(BackendWorkflowInitializationActor.scala:171)
	at cromwell.backend.sfs.SharedFileSystemInitializationActor.initSequence(SharedFileSystemInitializationActor.scala:37)
	at cromwell.backend.BackendWorkflowInitializationActor$$anonfun$receive$1$$anonfun$applyOrElse$4.apply(BackendWorkflowInitializationActor.scala:163)
	at cromwell.backend.BackendWorkflowInitializationActor$$anonfun$receive$1$$anonfun$applyOrElse$4.apply(BackendWorkflowInitializationActor.scala:163)
	at cromwell.backend.BackendLifecycleActor$class.performActionThenRespond(BackendLifecycleActor.scala:44)
	at cromwell.backend.sfs.SharedFileSystemInitializationActor.performActionThenRespond(SharedFileSystemInitializationActor.scala:37)
	at cromwell.backend.BackendWorkflowInitializationActor$$anonfun$receive$1.applyOrElse(BackendWorkflowInitializationActor.scala:163)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at cromwell.backend.sfs.SharedFileSystemInitializationActor.aroundReceive(SharedFileSystemInitializationActor.scala:37)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I guess it's something to do with 85216837ce62399b33c2b4d51c314d0bd98fe89a

### PR DESCRIPTION
**Pull request by ruchim**

### CODE REVIEW

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/DeclarationValidation.scala [hunk 1]
```diff
@@ -33,16 +33,19 @@ object DeclarationValidation {
...
-          case WdlStringType => new StringRuntimeAttributesValidation(declaration.unqualifiedName)
-          case other => throw new RuntimeException(s"Unsupported config runtime attribute $other ${declaration.unqualifiedName}")
-        }
-        new DeclarationValidation(declaration, validator)
+        val validatedRuntimeAttr = validator(declaration.wdlType, declaration.unqualifiedName)
+        new DeclarationValidation(declaration, validatedRuntimeAttr)
     }
   }
+
+  def validator(wdlType: WdlType, unqualifiedName: String): PrimitiveRuntimeAttributesValidation[_] = wdlType match {
```
Comment (mcovarr) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/DeclarationValidation.scala]: If this is going to be top-level it should be `private`

**Comment by mcovarr** [comment]
:+1:

**Comment by Horneth** [comment]
Is there an equivalent for JES runtime attributes validation that could need an update as well ?

**Comment by Horneth** [comment]
:+1:

**Comment by kshakir** [comment]
@Horneth 

> Is there an equivalent for JES runtime attributes validation that could need an update as well ?

Not that I can think of. JES's runtime attributes are [hardcoded](https://github.com/broadinstitute/cromwell/blob/23/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesRuntimeAttributes.scala#L19-L28) into the scala code. Meanwhile, these are the declarations of runtime attributes for the Config based backend. Via the config, one can specify the list of valid runtime attributes for _your_ backend, PBS, LSF, SGE, etc. See #1737 for an example of where this was broken.

File: .travis.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@ jdk:
   - oraclejdk8
 env:
   global:
-    - CENTAUR_BRANCH=develop
+    - CENTAUR_BRANCH=rm_moreSleeping
```
Comment (cjllanwarne) [inline .travis.yml]: Don't forget to switch this back before merging

**Comment by cjllanwarne** [comment]
Also for develop?

**Pull request merged**